recipe-client-addon/lib/CleanupManager.jsm

```diff
@@ -5,26 +5,49 @@
 "use strict";
 
 const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "AsyncShutdown", "resource://gre/modules/AsyncShutdown.jsm");
+
 this.EXPORTED_SYMBOLS = ["CleanupManager"];
 
-const cleanupHandlers = new Set();
+class CleanupManagerClass {
+  constructor() {
+    this.handlers = new Set();
+    this.cleanupPromise = null;
+  }
 
-this.CleanupManager = {
   addCleanupHandler(handler) {
-    cleanupHandlers.add(handler);
-  },
+    this.handlers.add(handler);
+  }
 
   removeCleanupHandler(handler) {
-    cleanupHandlers.delete(handler);
-  },
-
-  cleanup() {
-    for (const handler of cleanupHandlers) {
-      try {
-        handler();
-      } catch (ex) {
-        Cu.reportError(ex);
-      }
+    this.handlers.delete(handler);
+  }
+
+  async cleanup() {
+    if (this.cleanupPromise === null) {
+      this.cleanupPromise = (async () => {
+        for (const handler of this.handlers) {
+          try {
+            await handler();
+          } catch (ex) {
+            // TODO: Log error in a way that persists after shutdown
```

Comment:
I can't figure out a good way to log an error during shutdown that won't just disappear since the app is closing. Any ideas?

Reply:
I think most logs like this will show up a terminal if Firefox is run from there, so this might not be an issue. Alternatively you could do something complex like jamming log lines in a preference and log them at startup.

Reply:
Yeah, I wouldn't worry too much about this.